fontbe/src/cpal.rs

```diff
@@ -0,0 +1,87 @@
+//! Generates a [cpal](https://learn.microsoft.com/en-us/typography/opentype/spec/cpal) table.
+
+use crate::{
+    error::Error,
+    orchestration::{AnyWorkId, BeWork, Context, WorkId},
+};
+use fontdrasil::orchestration::{Access, Work};
+use fontir::orchestration::WorkId as FeWorkId;
+use log::debug;
+use write_fonts::{tables::cpal::ColorRecord, tables::cpal::Cpal, NullableOffsetMarker};
+
+#[derive(Debug)]
+struct CpalWork {}
+
+pub fn create_cpal_work() -> Box<BeWork> {
+    Box::new(CpalWork {})
+}
+
+impl Work<Context, AnyWorkId, Error> for CpalWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Cpal.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        Access::Variant(AnyWorkId::Fe(FeWorkId::ColorPalettes))
+    }
+
+    /// Generate [cpal](https://learn.microsoft.com/en-us/typography/opentype/spec/cpal)
+    fn exec(&self, context: &Context) -> Result<(), Error> {
+        let colors = context.ir.colors.try_get();
+        // Guard clause: no colors
+        if !colors
+            .as_ref()
+            .map(|c| c.palettes.iter().any(|p| !p.is_empty()))
+            .unwrap_or_default()
+        {
+            debug!("Skip cpal; no colors");
+            return Ok(());
+        }
+        let colors = colors.unwrap();
+        let sz0 = colors.palettes[0].len();
+        if colors.palettes.iter().any(|p| p.len() != sz0) {
+            return Err(Error::InconsistentPaletteLength(
+                colors.palettes.iter().map(Vec::len).collect(),
+            ));
+        }
```

Comment:
ditto here.. I would prefer to enforce our invariants earlier on, when we're constructing the palettes, instead of needing to handle the various edge cases here at the end.

Reply:
That's fair, will migrate it elsewhere.